styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -126,37 +136,120 @@
-        final Secret secret = client.secrets().withName(specSecret.name()).get();
-        if (secret == null) {
-          LOG.error("[AUDIT] Workflow {} refers to a non-existent secret {}",
-                    workflowInstance.workflowId(), specSecret.name());
-          throw new InvalidExecutionException(
-              "Referenced secret '" + specSecret.name() + "' was not found");
-        } else {
-          LOG.info("[AUDIT] Workflow {} refers to secret {}",
-                   workflowInstance.workflowId(), specSecret.name());
-        }
-      });
-
       final Pod pod = client.pods().create(createPod(workflowInstance, runSpec));
       return pod.getMetadata().getName();
     } catch (KubernetesClientException kce) {
       throw new IOException("Failed to create Kubernetes pod", kce);
     }
   }
 
+  private void ensureSecret(WorkflowInstance workflowInstance, RunSpec runSpec) {
+    if (runSpec.serviceAccount().isPresent()) {
+      final String secretName = buildSecretName(runSpec.serviceAccount().get());
+      final Secret secret = client.secrets().withName(secretName).get();
+      if (secret == null) {
+        final String serviceAccount = runSpec.serviceAccount().get();
+
+        // TODO: create keys calling GCP API, base64 already
+        final String jsonKey = "";
+        final String p12Key = "";
+
+        final Map<String, String> keys = Maps.newHashMap();
+        keys.put(STYX_WORKFLOW_SA_JSON_KEY, jsonKey);
+        keys.put(STYX_MANAGED_WORKFLOW_SA_P12_KEY, p12Key);
+        final Map<String, String> annotations = Maps.newHashMap();
+        // TODO: add key names returned by GCP API as annotations
+        annotations.put(STYX_WORKFLOW_SA_SECRET_ANNOTATION, serviceAccount);
+        client.secrets().create(new SecretBuilder()
+                                    .withNewMetadata()
+                                    .withName(secretName)
+                                    .withAnnotations(annotations)
+                                    .endMetadata()
+                                    .withData(keys)
+                                    .build());
+        LOG.info("[AUDIT] Secret {} created for service account {}", secretName, serviceAccount);
+      }
+    } else if (runSpec.secret().isPresent()) {
+      final WorkflowConfiguration.Secret specSecret = runSpec.secret().get();
+      final Secret secret = client.secrets().withName(specSecret.name()).get();
+      if (secret == null) {
+        LOG.error("[AUDIT] Workflow {} refers to a non-existent secret {}",
+                  workflowInstance.workflowId(), specSecret.name());
+        throw new InvalidExecutionException(
+            "Referenced secret '" + specSecret.name() + "' was not found");
+      } else {
+        LOG.info("[AUDIT] Workflow {} refers to secret {}",
+                 workflowInstance.workflowId(), specSecret.name());
+      }
+    }
+  }
+
+  private static String buildSecretName(String serviceAccount) {
```

Comment:
Do we want store replicated secrets? Now we have one `service_account` mapped to one `kube secret`. By prefixing component and workflow, we will end up having many secrets having keys pointing to the same SA.